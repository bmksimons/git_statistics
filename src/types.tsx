export interface Issue{
    title: string;
    state: string;
}

export interface Api_commits {
    id: string;
    short_id: string;
    created_at: string;
    parent_ids: string[];
    title: string;
    message: string;
    author_name: string;
    author_email: string;
    authored_date: string;
    committer_name: string;
    committer_email: string;
    committed_date: string;
    web_url: string;
  }

  export interface commitsByDate {
      [key: string] : number;
  }
