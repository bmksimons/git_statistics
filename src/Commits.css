.commits {
  /* border-color: black; */
  /* border-style: solid; */
  border-width: 5px;
  border-radius: 10px;
  margin: 10px;
}

.chart {
  margin: 10px;
}

.tooltip {
  background-color: rgba(136, 132, 216, 0.2);
  border: 2px solid black;
  border-radius: 5px;
  padding-left: 5px;
  padding-right: 5px;
}

.numCommits {
  margin: 0;
}

.date {
  margin: 0;
  margin-top: 3px;
  margin-bottom: 3px;
}
