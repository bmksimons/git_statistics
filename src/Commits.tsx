import { GITLAB_TOKEN } from "./tokens";
import React, { useEffect, useState } from "react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";
import { type } from "os";
import { stringify } from "querystring";

const url = "https://gitlab.stud.idi.ntnu.no/api/v4/projects/5834";
//const url = "https://gitlab.stud.idi.ntnu.no/api/v4/projects/11800";

interface api_commits {
  id: string;
  short_id: string;
  created_at: string;
  parent_ids: string[];
  title: string;
  message: string;
  author_name: string;
  author_email: string;
  authored_date: string;
  committer_name: string;
  committer_email: string;
  committed_date: string;
  web_url: string;
}

// i'nterface num_commits {
//   date: string;
//   numCommits: number;
// }'

export default function Commits() {
  const [commits, changeCommits] = useState<any[]>([]); // Probably useless now

  // list of object containing date and number of commits that date
  // that will be passed to the chart for ploting
  const [commitsData, changeCommitsData] = useState<{ [key: string]: number }>({
    "2021-09-14": 1,
  });

  // Dummy data for testing ploting -v
  const data = [
    {
      name: "Page A",
      commits: 2400,
    },
    {
      name: "Page B",
      commits: 1398,
    },
    {
      name: "Page C",
      commits: 9800,
    },
    {
      name: "Page D",
      commits: 3908,
    },
    {
      name: "Page E",
      commits: 4800,
    },
    {
      name: "Page F",
      commits: 3800,
    },
    {
      name: "Page G",
      commits: 4300,
    },
  ];

  async function getCommits() {
    try {
      const res = await fetch(url + "/repository/commits", {
        headers: new Headers({
          Authorization: "Bearer " + GITLAB_TOKEN,
        }),
      })
        return res.json();
          //console.log(data); // Loging only for testing purposes
          // changeCommits(data);
          //updateCommitData(data);
    } catch {
      console.log("Failed fetching commits");
    }
  }

  function updateCommitData(commits: api_commits[]) {
    for (let commit of commits) {
      const date = commit.committed_date.slice(0, 10);
      console.log(date)
      if (date in commitsData) {
        changeCommitsData((prevCommitsData) => {
          return { ...prevCommitsData, date: (prevCommitsData[date] += 1) };
        });
      } else {
        // This part works
        changeCommitsData((prevCommitsData) => {
          return { ...prevCommitsData, [date]: 1 };
        });
      }
    }
  }

  useEffect(() => {
    const fetchCommits = async () => {
      const commits = await getCommits();
      updateCommitData(commits);
    }
    fetchCommits();
  })

  return (
    <div>
      <p></p>
      <Chart data={data} />
    </div>
  );
}

//function presentCommits() {
//  getCommits().then((commits) => {
//    console.log(commits);
//  });
//}

function Chart(props: { data: any[] }) {
  // Source: https://recharts.org/en-US/api/LineChart
  // Source: https://recharts.org/en-US/examples/SimpleLineChart

  return (
    <LineChart
      width={500}
      height={300}
      data={props.data}
      margin={{
        top: 5,
        right: 30,
        left: 20,
        bottom: 5,
      }}
    >
      <CartesianGrid strokeDasharray="3 3" />
      <XAxis dataKey="name" />
      <YAxis />
      <Tooltip />
      <Legend />
      <Line
        type="monotone"
        dataKey="commits"
        stroke="#8884d8"
        activeDot={{ r: 8 }}
      />
    </LineChart>
  );
}
