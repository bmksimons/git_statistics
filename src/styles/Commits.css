.commits {
    margin: 5% 5%;
}