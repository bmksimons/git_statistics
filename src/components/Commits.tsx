import { useEffect, useState } from "react";
import { getCommitsFromGitlab } from "../api/ApisCalls";
import "../styles/Commits.css";
import { Api_commits, commitsByDate } from "../types";
import Chart from "./CommitsChart";
import ContentDescription from "./ContentDescription";
import DateSlider from "./DateSlider";

function getDates(startDateStr: string, stopDateStr: string) {
  /**
   * Retuns an array of objects with all dates from start date untill
   * today with number of commits that day equal to 0
   *
   * @param startDateStr - First date in the array as a string on the
   * format "YYYY-MM-DD"
   *
   * @returns List of objects from start date untill today on the
   * format [{date: "2000-01-01", commits: 0}, ...]
   */
  let currentDate = new Date(startDateStr);
  const stopDate = new Date(stopDateStr);
  let allDates: commitsByDate[] = [];

  while (currentDate <= stopDate) {
    let date: commitsByDate = {
      date: currentDate.toISOString().slice(0, 10),
      commits: 0,
    };
    allDates.push(date);
    currentDate.setDate(currentDate.getDate() + 1);
  }
  return allDates;
}

export default function Commits() {
  // list of object containing date and number of commits that date
  // that will be passed to the chart for ploting
  const [chartsData, setChartsData] = useState<commitsByDate[]>([
    { date: "2021-09-10", commits: 0 },
  ]);

  function updateCommitData(apiCommits: Api_commits[]) {
    /**
     * Function taking api respons from commits and updating state
     * of chartsData
     *
     * @param commits - Resons from GitLab API. Array of Api_commits
     */
    const firstCommitDate = apiCommits.slice(-1)[0].committed_date.slice(0, 10);
    const lastCommitDate = apiCommits[0].committed_date.slice(0, 10);

    // Array of commitsByDate that is used to update state of chatsData
    let newChartsDataState: commitsByDate[] = getDates(
      firstCommitDate,
      lastCommitDate
    );

    for (let apiCommit of apiCommits) {
      const date = apiCommit.committed_date.slice(0, 10);
      newChartsDataState = newChartsDataState.map((chartDataObject) => {
        // Increment number of commits that day if the chartDataObject
        // is equal to the date of the commit returned from the api
        if (chartDataObject.date === date) {
          return {
            ...chartDataObject,
            commits: chartDataObject.commits + 1,
          };
        }
        return chartDataObject;
      });
    }
    setChartsData(newChartsDataState);
  }

  useEffect(() => {
    const fetchCommits = async () => {
      try {
        const commits = await getCommitsFromGitlab();
        updateCommitData(commits);
      } catch (e) {
        console.log(e);
      }
    };
    fetchCommits();
  }, []);

  const props = {
    data: chartsData,
  };

  const contentDescriptionProps = {
    header: "Commits",
    content:
      "The graph below shows commits per day over the period chosen on the slider under the graph. 10th of september was the day of the first commits, and the end date on the graph is the date of the latest commit to the repository.",
  };

  return (
    <div className="commits">
      <ContentDescription {...contentDescriptionProps} />
      <Chart {...props} />
      <DateSlider {...props} />
    </div>
  );
}
