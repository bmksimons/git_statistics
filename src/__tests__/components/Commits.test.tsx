import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import Commits from '../../components/Commits'
import Chart from '../../components/CommitsChart'
import { commitsByDate } from "../../types";

test('renders commits', () => {
  const { container } = render(<Commits/>);
  expect(container).toMatchSnapshot();
  });