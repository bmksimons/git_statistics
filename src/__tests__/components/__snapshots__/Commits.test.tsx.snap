// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders commits 1`] = `
<div>
  <div
    class="commits"
  >
    <p />
    <div
      class="chart"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px;"
      />
    </div>
  </div>
</div>
`;
